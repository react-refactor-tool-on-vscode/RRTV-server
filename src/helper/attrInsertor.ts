export function addTabStop(text: string, option: number): string | undefined {
    if (!hasJSXCode(text)) return;
    let modifiedCode: string = text;
    const regex = /<(\w+)((?:\s+\w+(?:\s*=\s*(?:"(?:\\.|[^\\"])*"|'(?:\\.|[^\\'])*'|\{(?:\\.|[^\\}]|\n)*\}|[^'">\s]+))?)+\s*|\s*)(\/)?>/g;
    if (option === 1) {
        modifiedCode = text.replace(regex, (match, tagName, attributes, selfClosing) => {
            const attributeString = ' ${1:key}=${2:"value"}';
            const closingSlash = selfClosing ? '/' : '';
            return `<${tagName}${attributes}${attributeString}${closingSlash}>`
        });
    } else if (option === 2) {
        let i = 2;
        modifiedCode = text.replace(regex, (match, tagName, attributes, selfClosing) => {
            const attributeString = ` $\{1:key}=$\{${i}:"value"}`;
            i++;
            const closingSlash = selfClosing ? '/' : '';
            return `<${tagName}${attributes}${attributeString}${closingSlash}>`
        });
    } else if (option === 3) {
        let i = 2;
        modifiedCode = text.replace(regex, (match, tagName, attributes, selfClosing) => {
            const attributeString = ` $\{${i}:key}=$\{1:"value"}`;
            i++;
            const closingSlash = selfClosing ? '/' : '';
            return `<${tagName}${attributes}${attributeString}${closingSlash}>`
        });
    } else {
        let i = 1;
        modifiedCode = text.replace(regex, (match, tagName, attributes, selfClosing) => {
            const attributeString = ` $\{${i}:key}=$\{${i + 1}:"value"}`;
            i += 2;
            const closingSlash = selfClosing ? '/' : '';
            return `<${tagName}${attributes}${attributeString}${closingSlash}>`
        });
    }

    return modifiedCode;
}
export function hasJSXCode(code: string): boolean {
    const regex = /<([a-zA-Z][\w.-]*)\b/g; // 匹配以大写字母开头的标签名称
    return regex.test(code);
}
