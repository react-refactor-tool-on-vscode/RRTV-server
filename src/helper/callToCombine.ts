import * as j from 'jscodeshift'
import { locToRange } from './locToRange';
export function transformer(text, index) {
    let root;
     try {
       root =  j(text)
    } catch (error) {
        return {
            check: false
        }
    };

    /// 找到函数声明
    const fd = root.find(j.FunctionDeclaration)
    const vd = root.find(j.VariableDeclaration)
    const declaration = fd.length ? fd : vd

    /// 建构名字列表
    const fd_namelist = new Set<any>()
    const jsx_namelist = new Set<any>()
    //.... 添加名字到 set 中

    /// 分类所有 jsx 元素 
    const je = root.findJSXElements()
    const c_je = {}
    je.forEach((path: any) => {
        const name = path.node.openingElement.name.name
        if (!c_je[name]) {
            c_je[name] = []
            jsx_namelist.add(name)
        }
        c_je[name].push(path)
    })

    /// 分类所有 声明
    const c_d = {}
    declaration.forEach(path => {
        const name = path.node.id.name
        if (!c_d[name]) {
            c_d[name] = []
            fd_namelist.add(name)
        }
        c_d[name].push(path)
    })

    /// 遍历名字列表
    // 为了保存 chain: {, ,}
    const chainlist = new Set<any>()
    // 开始遍历
console.log(fd_namelist)
    fd_namelist.forEach((name1: string) => {
        console.log(c_d[name1])
        j(c_d[name1]).findJSXElements().paths()
            .map((e: any) => {
                return e.node.openingElement.name.name
            })
            .filter(name => {
                return name !== undefined
            })
            .forEach(name2 => {
                if (!(name2 in c_d))
                    return
                const jsxnames = j(c_d[name2]).findJSXElements().paths()
                    .map((e: any) => {
                        return e.node.openingElement.name.name
                    })
                    .filter(name => name !== undefined)
                    .forEach(name3 => {
                        if (!(name3 in c_d))
                            return
                        const newchain = name1 + '.' + name2 + '.' + name3
                        chainlist.add(newchain)
                    })
            })
    })
    console.log(chainlist, "list")
    if (chainlist.size !== 1)
        return { check: false }
    const [App, Parent, Child] = Array.from(chainlist)[0].split('.')

    /// 获得 "App" "Parent" "Child"
    const d_app = c_d[App]
    const d_parent = c_d[Parent]
    const d_child = c_d[Child]

    const j_app = c_je[App]
    const j_parent = c_je[Parent]
    const j_child = c_je[Child]
    /// 判断与构建 range
    let checkIndex = false
    const r_app = d_app.map(path => { if (handleIndex(index, path)) checkIndex = true; return locToRange(path.node.loc) })
    const r_parent = d_parent.map(path => { if (handleIndex(index, path)) checkIndex = true; return locToRange(path.node.loc) })
    const r_child = d_child.map(path => { if (handleIndex(index, path)) checkIndex = true; return locToRange(path.node.loc) })
    const range = [
        r_app,
        r_parent,
        r_child
    ]

    if (!checkIndex) 
        return { check: false }
    /// 更改 Parent declaration [replaceWith]
    const pattern = j(d_parent[0]).find(j.ObjectPattern).paths()[0]
    const new_property = j.property("init", j.identifier("children"), j.identifier("children"))
    new_property.shorthand = true
    const properties = pattern.node.properties
    properties.push(new_property)

    const return_jsx = j(d_parent[0]).find(j.JSXElement)
    const new_args = pattern.node.properties.filter((node: any) => node.value.name !== "children").map((node: any) => node.key)
    const new_return = j.callExpression(j.identifier("children"), new_args)
    return_jsx.replaceWith(new_return)

    /// 更改 App declaration -> return
    const parent_element = j(d_app).findJSXElements(Parent)

    // 生成 {text => ...
    const expression = j.arrowFunctionExpression(
        new_args,
        j.jsxElement(
            j.jsxOpeningElement(
                j.jsxIdentifier(Child),
                new_args.map(arg => j.jsxAttribute(j.jsxIdentifier(arg.name), j.jsxExpressionContainer(arg))),
                true
            ),
            null,
            []
        )
    );
    console.log(expression)

    const op_element = parent_element.paths()
        .map(path => path.node.openingElement)
    const new_closeElement = j.jsxClosingElement(j.jsxIdentifier(Parent))
    op_element.forEach(node => node.selfClosing = false)
    const new_jsxElement = op_element.map(e => j.jsxElement(e, new_closeElement, [j.jsxText("\n"), j.jsxExpressionContainer(expression), j.jsxText("\n")]))
    for (let i = 0; i < new_jsxElement.length; ++i) {
        parent_element.at(i).replaceWith(new_jsxElement[i])
    }

    return { check: true, newText: root.toSource(), newRange: range };
}

function handleIndex(index: number, path: any): boolean {
    // seePosition(path, index)
    if (path.node.start <= index && index <= path.node.end) {
        return true
    }
    return false
}
// const res: boolean | string = transformer(text)
// if (typeof res === "boolean" && !res) 没有调用到组合的情境
// else res.newText, 替换全文
// res.newRange.forEach(range=>...) 

const text = `function CardBoard() {
	return (<>
		<Card text="text1" id="3" />
		<Card text="text2" id="4" />
		<Card text="text3" id="5" />
	</>
	)
}

function Card({text, id}) {
	return <CardDrawer text={text} id={id}/>
}

function CardDrawer({text, id}) {
	return <div id={id}>{text}</div>
}`

const index = 3

console.log(transformer(text, index))