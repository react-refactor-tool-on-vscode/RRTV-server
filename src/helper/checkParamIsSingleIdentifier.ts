import { Range } from "vscode-languageserver";
import parseToAst from "./ParseToAst";
import traverse from "@babel/traverse";
import * as t from "@babel/types";
import { IsRangeInLoc } from "./RangeLoc";

export function checkParamIsSingleIdentifier(code: string, range: Range): boolean {
    let result = false;
    const ast = parseToAst(code);
    traverse(ast, {
        enter(path) {
            if (t.isFunctionDeclaration(path.node) &&
                IsRangeInLoc(range, path.node.loc)) {
                const param = path.node.params[0];
                if (t.isIdentifier(param) && path.node.params.length == 1) {
                    result = true;
                }
            }
        },
    });
    return result;
}
