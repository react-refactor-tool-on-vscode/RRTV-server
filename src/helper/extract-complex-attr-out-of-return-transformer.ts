/* 
// ============= Example ================
const text = `const renderButton = (text, onClick) => {
    return (
        <>
            <button onClick={() => {
                console.log('Button clicked!');
            }}>{text}</button>
            <button onClick={function () {
                console.log('Button clicked!')
            }}></button>
            <button onClick={newFunction}></button>
            <button onClick={function (e) {
                console.log('Button clicked!', e)
            }}></button>
            <button onClick={e => NewFunction(e)}></button>
            <button onClick={function (e, ee) {
                console.log('Button clicked!', ee, e)
            }}></button>
            <button onClick={(e) => {
                console.log('Button clicked!', e);
            }}>{text}</button>
            <button onClick={function(e) {
                return NewFunction00(e)
            }}></button>
        </>
    );
}`
const index = 120 */

import * as jscodeshift from 'jscodeshift';

const transform = (file: jscodeshift.FileInfo, api: jscodeshift.API, options: jscodeshift.Options) => {
    const j = api.jscodeshift;
    let root;
    try {
        root = j(file.source);
    } catch (error) {
        return {
            newText: undefined,
            newRange: undefined
        }
    }
    let count = 0
    let _handleType = new handleType()
    root.find(j.ArrowFunctionExpression)
        .filter(path => _handleType.arrowFunctionExpression(path) && handleIndex(options.index, path))
        .forEach((path) => { handler(j, path, count++, _handleType.args) })
    root.find(j.FunctionExpression)
        .filter(path => _handleType.functinoExpression(path, j, _handleType.args) && handleIndex(options.index, path))
        .forEach((path) => { handler(j, path, count++, _handleType.args) })

    if (_handleType.args[0]) {
        const newRange = _handleType.args[0].get(0).node.loc
        return {
            newText: root.toSource(),
            newRange
        }
    }
    else {
        return {
            newText: undefined,
            newRange: undefined
        }
    }
}

function handler(j: jscodeshift.JSCodeshift, path: any, count: number, ...rest: any[]) {
    const jsxAttributes = j(path).closest(j.JSXAttribute)
    if (!jsxAttributes.length) return
    const functionName = "NewFunction" + count
    const newFunction = j.functionDeclaration(
        j.identifier("NewFunction" + count),
        path.node.params,
        j.blockStatement(path.node.body.body)
    );

    const returns = jsxAttributes.closest(j.ReturnStatement)
    returns.insertBefore(newFunction)

    const newFunctionCall = j.callExpression(j.identifier(functionName), path.node.params);
    const newArrowFunctionExpression = j.arrowFunctionExpression(path.node.params, newFunctionCall)
    if (path.node.params.length === 0)
        j(path).replaceWith(j.identifier(functionName)) //ok
    else
        j(path).replaceWith(newArrowFunctionExpression)

    const findFD = returns.closest(j.FunctionDeclaration)
    const findDefination = findFD.length ? findFD : returns.closest(j.VariableDeclaration);
    rest[0].push(findDefination)
}

function handleIndex(index: number, path: any): boolean {
    // seePosition(path, index)
    if (path.node.start <= index && index <= path.node.end) {
        return true
    }
    return false
}

function seePosition(path: any, ...rest: any[]) {
    console.log(path.node.start, path.node.end, ...rest)
}

class handleType {
    constructor() { }
    public args: any[] = []
    arrowFunctionExpression(path: any, ...rest: any[]) {
        /// 不识别 <button onClick={e => NewFunction(e)}></button>
        const body = path.node.body;
        return (
            body.type === "JSXElement" ||
            body.type === "FunctionExpression" ||
            body.type === "ArrowFunctionExpression" ||
            body.type === "BlockStatement"
        );
    }
    functinoExpression(path: any, ...rest: any[]) {
        /// 不识别 <button onClick={function(e) { return NewFunction(e)}}></button>
        const body = path.node.body;
        /// j 放在 rest[0]
        let check = false
        if (rest.length) {
            // console.log(rest[0](path).find(rest[0].ReturnStatement).length ? rest[0](path).find(rest[0].ReturnStatement).get(0).node.argument.type : null) /// true
            check = rest[0](path).find(rest[0].ReturnStatement, {
                argument: {
                    type: "CallExpression"
                }
            }).length > 0
            // console.log(check)
        }
        return !check
    }
}
export const checkAttributeExtract = (text: string, index: number) => {
    const output = transform(
        {
            source: text,
            path: ""
        },
        {
            jscodeshift,
            j: jscodeshift,
            stats: () => { },
            report(msg) {

            },
        },
        {
            index
        }
    )
    return output;

    // console.log(output)
}

/* const output = checkAttributeExtract(text, index) */

/* if (output.newText === undefined) {
    console.error("条件不满足")
}
else {
    console.log("条件满足")
    /// use output.newText & output.newRange
    const newRange = {
        start: {
            line: output.newRange.start.line,
            character: output.newRange.start.column
        },
        end: {
            line: output.newRange.end.line,
            character: output.newRange.end.column
        }
    }
} */
