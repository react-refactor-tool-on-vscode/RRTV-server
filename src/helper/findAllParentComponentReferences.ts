import { Range } from "vscode-languageserver";
import parseToAst from "./ParseToAst";
import traverse, { NodePath, TraverseOptions } from "@babel/traverse";
import * as t from "@babel/types";
import { IsRangeInLoc } from "./RangeLoc";

export function findAllParentComponentReferences(
    code: string,
    range: Range
): Map<t.SourceLocation, t.SourceLocation[]> {
    const ast = parseToAst(code);
    const result = new Map<t.SourceLocation, t.SourceLocation[]>();
    let idsForFunction = [];
    let parentComponentLoc: t.SourceLocation;
    let liftedName: string;
    traverse(ast, {
        enter(path) {
            if (
                t.isFunctionDeclaration(path.node) &&
                IsRangeInLoc(range, path.node.loc)
            ) {
                liftedName = path.node.id.name;
            }
        },
    });

    const findAllRefs: TraverseOptions = {
        JSXIdentifier(path) {
            if (path.node.name == liftedName) {
                idsForFunction.push(path.node.loc);
            }
        },
    };

    traverse(ast, {
        enter(path) {
            if (t.isFunctionDeclaration(path.node)) {
                parentComponentLoc = path.node.id.loc;
                (path.get("body") as NodePath<t.Node>).traverse(findAllRefs);
                if (idsForFunction.length != 0)
                    result.set(parentComponentLoc, idsForFunction);
                idsForFunction = [];
            }
        },
    });
    return result;
}
