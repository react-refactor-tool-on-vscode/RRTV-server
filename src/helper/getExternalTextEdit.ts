import { Range, TextEdit } from "vscode-languageserver";
import * as t from "@babel/types";
import traverse from "@babel/traverse";
import { findFunctionDeclarationOfRange } from "./findFunctionDeclarationOfRange";
import { isHookCallExpression } from "./isHookCallExpression";
import {
    locEndToPosition,
    locStartAfterToPosition,
    locToRange,
} from "./locToRange";
import generate from "@babel/generator";
import * as _ from "lodash";

export function getExternalTextEdit(
    isIntoNewComponent: boolean,
    refs: Map<t.SourceLocation, t.SourceLocation[]>,
    ast: t.Node,
    range: Range,
    newId?: string
): TextEdit[] {
    const result: TextEdit[] = [];
    let callExpressionNode: t.CallExpression;
    const liftedComponent = findFunctionDeclarationOfRange(range, ast);
    traverse(ast, {
        CallExpression(path) {
            if (isHookCallExpression(range, path)) {
                callExpressionNode = path.node;
            }
        },
    });
    const newStateNode = t.variableDeclaration("const", [
        t.variableDeclarator(t.identifier("state"), callExpressionNode),
    ]);
    const newStateAttr = t.jsxAttribute(
        t.jsxIdentifier("state"),
        t.jsxExpressionContainer(t.identifier("state"))
    );
    if (isIntoNewComponent) {
        // Create new component
        const newComponentNode = t.functionDeclaration(
            t.identifier(newId!),
            [t.identifier("props")],
            t.blockStatement([
                newStateNode,
                t.returnStatement(
                    t.jsxElement(
                        t.jsxOpeningElement(
                            t.jsxIdentifier(liftedComponent.node.id.name),
                            [
                                t.jsxSpreadAttribute(t.identifier("props")),
                                newStateAttr,
                            ],
                            true
                        ),
                        null,
                        []
                    )
                ),
            ])
        );
        result.push(
            TextEdit.insert(
                locEndToPosition(liftedComponent.node.loc),
                generate(newComponentNode).code + "\n"
            )
        );
        // Traverse refs and change identifier
        for (const value of refs.values()) {
            value.forEach((idLoc) => {
                result.push(TextEdit.replace(locToRange(idLoc), newId));
            });
        }
    } else {
        // Traverse parent component: add state and modify jsx.
        for (const [key, value] of refs) {
            traverse(ast, {
                FunctionDeclaration(path) {
                    if (
                        key.start.line == path.node.id.loc.start.line &&
                        key.start.column == path.node.id.loc.start.column &&
                        key.end.line == path.node.id.loc.end.line &&
                        key.end.column == path.node.id.loc.end.column
                    ) {
                        const blockStartLoc = path.node.body.loc;
                        result.push(
                            TextEdit.insert(
                                locStartAfterToPosition(blockStartLoc),
                                "\n" + generate(newStateNode).code
                            )
                        );
                    }
                },
            });
            value.forEach((jsxIdentifierLoc) => {
                result.push(
                    TextEdit.insert(
                        locEndToPosition(jsxIdentifierLoc),
                        "\n" + generate(newStateAttr).code + "\n"
                    )
                );
            });
        }
    }
    return result;
}
