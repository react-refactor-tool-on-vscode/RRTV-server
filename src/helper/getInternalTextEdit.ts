import { Range, TextEdit } from "vscode-languageserver";
import * as t from "@babel/types";
import traverse from "@babel/traverse";
import { IsRangeInLoc } from "./RangeLoc";
import { locToRange } from "./locToRange";
import generate from "@babel/generator";
import { isHookCallExpression } from "./isHookCallExpression";


export function getInternalTextEdit(
    isParamSingleId: boolean,
    ast: t.Node,
    range: Range
): TextEdit[] {
    const result: TextEdit[] = [];

    let singleIdParamName: string;

    traverse(ast, {
        enter(path) {
            if (t.isFunction(path.node) && IsRangeInLoc(range, path.node.loc)) {
                if (isParamSingleId) {
                    if (path.node.params.length == 1 &&
                        t.isIdentifier(path.node.params[0])) {
                        singleIdParamName = path.node.params[0].name;
                    }
                } else {
                    const isRightParam = path.node.params.length == 1 &&
                        t.isObjectPattern(path.node.params[0]);
                    if (isRightParam) {
                        const param = path.node.params[0] as t.ObjectPattern;
                        param.properties.push(
                            t.objectProperty(
                                t.identifier("state"),
                                t.identifier("state")
                            )
                        );
                        result.push(
                            TextEdit.replace(
                                locToRange(path.node.params[0].loc),
                                generate(param).code
                            )
                        );
                    }
                }
            }
        },
    });

    traverse(ast, {
        CallExpression(path) {
            if (isHookCallExpression(range, path)) {
                if (isParamSingleId) {
                    result.push(
                        TextEdit.replace(
                            locToRange(path.node.loc),
                            generate(
                                t.memberExpression(
                                    t.identifier(singleIdParamName),
                                    t.identifier("state")
                                )
                            ).code
                        )
                    );
                } else {
                    result.push(
                        TextEdit.replace(
                            locToRange(path.node.loc),
                            generate(t.identifier("state")).code
                        )
                    );
                }
            }
        },
    });

    return result;
}
