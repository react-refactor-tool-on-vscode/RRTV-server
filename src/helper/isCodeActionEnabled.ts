import { Range } from "vscode-languageserver";
import parseToAst from "./ParseToAst";
import traverse from "@babel/traverse";
import * as t from "@babel/types";
import { IsRangeInLoc } from "./RangeLoc";

export function isCodeActionEnabled(code: string, range: Range): boolean {
    const ast = parseToAst(code);
    let result = false;
    traverse(ast, {
        CallExpression(path) {
            if (t.isIdentifier(path.node.callee) &&
                IsRangeInLoc(range, path.node.loc)) {
                if (path.node.callee.name.startsWith("use")) {
                    result = true;
                }
            }
        },
    });
    return result;
}
