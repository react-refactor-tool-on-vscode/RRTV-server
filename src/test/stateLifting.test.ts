import { Range } from "vscode-languageserver";
import { isCodeActionEnabled } from "../helper/isCodeActionEnabled";
import { findAllParentComponentReferences } from "../helper/findAllParentComponentReferences";
import { checkParamIsSingleIdentifier } from "../helper/checkParamIsSingleIdentifier";
import parseToAst from "../helper/ParseToAst";
import { getInternalTextEdit } from "../helper/getInternalTextEdit";
import { getExternalTextEdit } from "../helper/getExternalTextEdit";
import { SourceLocation } from "@babel/types";

test("check is code action enabled", () => {
    const usage1 = {
        code: `function Test({item}) {const [count, setCount] = useState(0);
            return <div onClick={() => setCount(count + 1)}>{item.text}</div>
        }`,
        range: Range.create(0, 52, 0, 54),
    };

    const result = isCodeActionEnabled(usage1.code, usage1.range);

    expect(result).toBe(true);

    const usage2 = {
        code: usage1.code,
        range: Range.create(0, 39, 0, 41),
    };

    const result2 = isCodeActionEnabled(usage2.code, usage2.range);

    expect(result2).toBe(false);
});

test("test find all parent component references", () => {
    const usage1 = {
        code: `function App(props) {const item = { text: "Hello?" }
            return <Test item={item} />
        }
        function Test({ item }) {
            const [count, setCount] = useState(0);
            return <div onClick={() => setCount(count + 1)}>{item.text}</div>
        }`,
        range: Range.create(4, 43, 4, 44),
    };

    const result = findAllParentComponentReferences(usage1.code, usage1.range);

    expect(result.size).toBe(1);

    console.log(123);
});

test("test check function param is single identifier", () => {
    const usage1 = {
        code: `function Test({item}) {const [count, setCount] = useState(0);
            return <div onClick={() => setCount(count + 1)}>{item.text}</div>
        }`,
        range: Range.create(0, 52, 0, 54),
    };

    const result1 = checkParamIsSingleIdentifier(usage1.code, usage1.range);

    expect(result1).toBe(false);

    const usage2 = {
        code: `function Test(item) {const [count, setCount] = useState(0);
            return <div onClick={() => setCount(count + 1)}>{item.text}</div>
        }`,
        range: Range.create(0, 52, 0, 54),
    };

    const result2 = checkParamIsSingleIdentifier(usage2.code, usage2.range);

    expect(result2).toBe(true);

    const usage3 = {
        code: `function Test(item,text) {const [count, setCount] = useState(0);
            return <div onClick={() => setCount(count + 1)}>{item.text}</div>
        }`,
        range: Range.create(0, 57, 0, 59),
    };
    const result3 = checkParamIsSingleIdentifier(usage3.code, usage3.range);

    expect(result3).toBe(false);
});

test("get internal test edit for single param", () => {
    const code1 = `function ComponentWithState(props) {
        const [state1, setState1] = useState(0);
    
        return (
            <div onclick={() => setState1(state1 + 1)}>
                <div>{props.text}</div>
                <span>{state1}</span>
            </div>)
    }`;

    const ast1 = parseToAst(code1);

    const result1 = getInternalTextEdit(true, ast1, Range.create(1, 40, 1, 42));
    expect(result1).toMatchSnapshot();
});

test("get internal test edit for single object pattern", () => {
    const code1 = `function ComponentWithState({item}) {
        const [state1, setState1] = useState(0);
    
        return (
            <div onclick={() => setState1(state1 + 1)}>
                <div>{item.text}</div>
                <span>{state1}</span>
            </div>)
    }`;

    const ast1 = parseToAst(code1);

    const result1 = getInternalTextEdit(
        false,
        ast1,
        Range.create(1, 40, 1, 42)
    );
    expect(result1).toMatchSnapshot();
});

test("get internal test edit for wrong param", () => {
    const code1 = `function ComponentWithState([a,b]) {const [count,setCount] = useState(0);return <div>{a}</div>};}`;
    const ast1 = parseToAst(code1);
    const result1 = getInternalTextEdit(true, ast1, Range.create(0, 67, 0, 67));
    expect(result1).toMatchSnapshot();
});

test("get external test edit without external refs", () => {
    const code1 = `function ComponentWithState({item}) {
        const [state1, setState1] = useState(0);
    
        return (
            <div onclick={() => setState1(state1 + 1)}>
                <div>{item.text}</div>
                <span>{state1}</span>
            </div>)
    }`;

    const ast1 = parseToAst(code1);

    const result1 = getExternalTextEdit(
        true,
        new Map<SourceLocation, SourceLocation[]>(),
        ast1,
        Range.create(1, 40, 1, 40),
        "StateContainer"
    );

    expect(result1).toMatchSnapshot();
});

test("get external test edit with external refs", () => {
    const code = `function ComponentWithState({item}) {
        const [state1, setState1] = useState(0);
    
        return (
            <div onclick={() => setState1(state1 + 1)}>
                <div>{item.text}</div>
                <span>{state1}</span>
            </div>)
    }
function App() {return <ComponentWithState item={{text="123"}} />}`;
    const ast = parseToAst(code);
    const range = Range.create(0, 41, 0, 41);

    const map = findAllParentComponentReferences(
        code,
        range
    );
    const result = getExternalTextEdit(true, map, ast, range, "StateContainer");

    expect(result).toMatchSnapshot();
});
