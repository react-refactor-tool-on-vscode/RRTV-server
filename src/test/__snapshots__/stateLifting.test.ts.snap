// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get external test edit with external refs 1`] = `
[
  {
    "newText": "function StateContainer(props) {
  const state;
  return <ComponentWithState {...props} state={state} />;
}
",
    "range": {
      "end": {
        "character": 5,
        "line": 8,
      },
      "start": {
        "character": 5,
        "line": 8,
      },
    },
  },
  {
    "newText": "StateContainer",
    "range": {
      "end": {
        "character": 42,
        "line": 9,
      },
      "start": {
        "character": 24,
        "line": 9,
      },
    },
  },
]
`;

exports[`get external test edit without external refs 1`] = `
[
  {
    "newText": "function StateContainer(props) {
  const state = useState(0);
  return <ComponentWithState {...props} state={state} />;
}
",
    "range": {
      "end": {
        "character": 5,
        "line": 8,
      },
      "start": {
        "character": 5,
        "line": 8,
      },
    },
  },
]
`;

exports[`get external text edit with external refs to existed parent component 1`] = `
[
  {
    "newText": "const state;",
    "range": {
      "end": {
        "character": 16,
        "line": 9,
      },
      "start": {
        "character": 16,
        "line": 9,
      },
    },
  },
  {
    "newText": "
state={state}",
    "range": {
      "end": {
        "character": 42,
        "line": 9,
      },
      "start": {
        "character": 42,
        "line": 9,
      },
    },
  },
]
`;

exports[`get internal test edit for single object pattern 1`] = `
[
  {
    "newText": "{
  item,
  state: state
}",
    "range": {
      "end": {
        "character": 34,
        "line": 0,
      },
      "start": {
        "character": 28,
        "line": 0,
      },
    },
  },
  {
    "newText": "state",
    "range": {
      "end": {
        "character": 47,
        "line": 1,
      },
      "start": {
        "character": 36,
        "line": 1,
      },
    },
  },
]
`;

exports[`get internal test edit for single param 1`] = `
[
  {
    "newText": "props.state",
    "range": {
      "end": {
        "character": 47,
        "line": 1,
      },
      "start": {
        "character": 36,
        "line": 1,
      },
    },
  },
]
`;

exports[`get internal test edit for wrong param 1`] = `[]`;
